### PR TITLE
Fix filter and in clause parsing

### PR DESCRIPTION
**Pull request by elonazoulay**
Fix issue with in clause in passthrough query.
TODO: fix functions in filter and add tests.

### CODE REVIEW

**Comment by findepi** [comment]
Is the fixed case testable?

**Comment by fx19880617** [comment]
Can you add a sample query supported by this PR?

**Comment by elonazoulay** [comment]
Updated, use PinotQuery to parse the filter. Added the ability to parse case statements and added tests.

**Comment by findepi** [comment]
@elonazoulay is this still "[WIP]"?

also, mind red CI result

**Comment by elonazoulay** [comment]
Thanks, will do @findepi ! Is it ok to change the pull request name or should I create a new one?

**Comment by findepi** [comment]
@elonazoulay change is fine

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/DynamicTableBuilder.java [hunk 1]
```diff
@@ -142,11 +142,6 @@ public static DynamicTable buildFromPql(PinotMetadata pinotMetadata, SchemaTable
 
     private static String getOutputColumnName(AggregationInfo aggregationInfo, String pinotColumnName)
     {
-        /*
-        if (pinotColumnName.equals(WILDCARD)) {
-            pinotColumnName = STAR;
-        }
-         */
```
Comment (martint) [inline plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/DynamicTableBuilder.java:149]: Move to a separate commit

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/DynamicTableBuilder.java [hunk 2]
```diff
@@ -32,6 +32,7 @@
 import java.util.Set;
```
Comment (martint) [inline plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/DynamicTableBuilder.java:32]: "Fix filter and in clause parsing" is a little vague. Can you add some details in the commit message explaining what was broken?

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/DynamicTablePqlExtractor.java [hunk 3]
```diff
@@ -115,4 +114,9 @@ private static String convertAggregationExpressionToPql(AggregationExpression ag
     {
         return format("%s(%s)", aggregationExpression.getAggregationType(), aggregationExpression.getBaseColumnName()).toLowerCase(ENGLISH);
     }
+
+    public static String encloseInParentheses(String value)
+    {
+        return format("(%s)", value);
+    }
```
Comment (martint) [inline plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/DynamicTablePqlExtractor.java:121]: Move this to a separate commit, as it seems unrelated to the main purpose of this commit.

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/FilterToPinotSqlConverter.java [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+    private FilterToPinotSqlConverter() {}
+
+    public static String convertFilter(org.apache.pinot.common.request.PinotQuery pinotQuery, Map<String, ColumnHandle> columnHandles)
+    {
+        StringBuilder builder = new StringBuilder();
+        new Visitor(builder, columnHandles).visitExpression(pinotQuery.getFilterExpression());
+        return builder.toString();
+    }
+
+    private static class Visitor
```
Comment (martint) [inline plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/FilterToPinotSqlConverter.java:57]: This is not a "visitor" in the traditional sense of the word.

Also, having a stateful class with side-effects on the underlying StringBuilder is error prone and confusing. I would restructure it as either:
* a set of static methods that take a StringBuilder and append the formatted string to it
* or better, have each method return a String and let the caller compose them, similar to how ExpressionFormatter works. This is the easiest form to reason about. This will also allow for much simpler forms using String.format(), Joiner, etc.

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/DynamicTableBuilder.java (hunk 2 above)
Comment (elonazoulay) [reply plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/DynamicTableBuilder.java:32]: Updated.

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/FilterToPinotSqlConverter.java (hunk 4 above)
Comment (elonazoulay) [reply plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/FilterToPinotSqlConverter.java]: Updated, lmk, also thanks for the advice!

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/FilterToPinotSqlConverter.java [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+
+        Expression right = functionCall.getOperands().get(1);
+        if (right.getType() != LITERAL || !formatLiteral(right.getLiteral()).equals("0")) {
+            return Optional.empty();
+        }
+        Function minus = left.getFunctionCall();
+        return Optional.of(format("(%s) = (%s)", formatExpression(minus.getOperands().get(0)), formatExpression(minus.getOperands().get(1))));
+    }
+
+    private String parseInClause(Function functionCall)
```
Comment (martint) [inline plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/FilterToPinotSqlConverter.java:121]: formatInClause?

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/FilterToPinotSqlConverter.java [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+            operator = "NOT IN";
+        }
+        return format("%s %s (%s)", formatExpression(functionCall.getOperands().get(0)),
+                operator,
+                functionCall.getOperands().subList(1, functionCall.getOperands().size()).stream()
+                        .map(this::formatExpression)
+                        .collect(joining(", ")));
+    }
+
+    private String parseCaseStatement(Function functionCall)
```
Comment (martint) [inline plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/FilterToPinotSqlConverter.java:141]: formatCaseStatement?

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/FilterToPinotSqlConverter.java [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+    {
+        checkState(functionCall.getOperator().equalsIgnoreCase("case"), "Unexpected operator '%s'", functionCall.getOperator());
+        checkState(functionCall.getOperands().size() >= 2, "Unexpected expression");
+        int whenStatements = functionCall.getOperands().size() / 2;
+        String caseStatement = format("CASE %s", IntStream.range(0, whenStatements)
+                .mapToObj(index -> "WHEN " + formatExpression(functionCall.getOperands().get(index)) +
+                        " THEN " + formatExpression(functionCall.getOperands().get(index + whenStatements)))
+                .collect(joining(" ")));
```
Comment (martint) [inline plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/FilterToPinotSqlConverter.java:149]: This is a little hard to read. It might be clearer to do it with a StringBuilder and a regular for loop

Comment (elonazoulay) [reply plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/FilterToPinotSqlConverter.java:149]: Updated

File: plugin/trino-pinot/src/test/java/io/trino/plugin/pinot/TestMinimalFunctionality.java [hunk 8]
```diff
@@ -80,7 +80,7 @@ protected QueryRunner createQueryRunner()
                 .build();
 
         return PinotQueryRunner.createPinotQueryRunner(
-                ImmutableMap.of(),
+                ImmutableMap.of("http-server.http.port", "8080"),
```
Comment (martint) [inline plugin/trino-pinot/src/test/java/io/trino/plugin/pinot/TestMinimalFunctionality.java:83]: Unrelated change?

Comment (elonazoulay) [reply plugin/trino-pinot/src/test/java/io/trino/plugin/pinot/TestMinimalFunctionality.java:83]: yep, forgot to remove while testing.
